autogpt_platform/backend/backend/blocks/google/gmail.py

```diff
@@ -772,24 +1035,81 @@ def _get_thread(self, service, thread_id: str, scopes: list[str] | None) -> Thre
         thread["messages"] = parsed_messages
         return thread
 
-    def _get_email_body(self, msg):
-        payload = msg.get("payload")
-        if not payload:
-            return "This email does not contain a text body."
-
-        if "parts" in payload:
-            for part in payload["parts"]:
-                if part.get("mimeType") == "text/plain" and "data" in part.get(
-                    "body", {}
-                ):
-                    return base64.urlsafe_b64decode(part["body"]["data"]).decode(
-                        "utf-8"
-                    )
-        elif payload.get("mimeType") == "text/plain" and "data" in payload.get(
-            "body", {}
-        ):
-            return base64.urlsafe_b64decode(payload["body"]["data"]).decode("utf-8")
-        return "This email does not contain a text body."
+    def _get_email_body(self, msg, service):
+        """Extract email body content with support for multipart messages and HTML conversion."""
+        text = self._walk_for_body(msg["payload"], msg["id"], service)
+        return text or "This email does not contain a readable body."
+
+    def _walk_for_body(self, part, msg_id, service, depth=0):
```

Comment:
https://github.com/Significant-Gravitas/AutoGPT/pull/10071#issuecomment-2931213476

Reply:
main reason is the extra effort needed for raw processing. It always returns the full body and attachments that *must* be dealt with. we can switch but i'd prefer not to deal with it right now